doc/man/kmousetool/kmousetool.1
.TH kmousetool 1

.SH NAME
kmousetool \- A TDE tool to help click the mouse.

.SH SYNOPSIS
.B kmousetool
.I [ Generic Options] [Qt-options] [TDE-options]

.SH DESCRIPTION
KMouseTool clicks the mouse whenever the mouse cursor pauses briefly.
It was designed to help those with repetitive strain injuries, for
whom pressing buttons hurts. It includes features and options that
provide artificial intelligence on common mouse gestures to perform
actions.

.PP
KMouseTool can also be configured to wait for a specified time for a
drag to begin, before unclicking the mouse. This way, you can use it for
drag and drop operations as well.

.SH GENERIC OPTIONS
.TP
.B \-\-help
Show help about options.
.TP
.B \-\-help-qt
Show Qt specific options.
.TP
.B \-\-help-tde
Show TDE specific options.
.TP
.B \-\-help-all
Show all options.
.TP
.B \-\-author
Show author information.
.TP
.B \-v, \-\-version
Show version information.

.SH QT OPTIONS
.TP
.B \-\-display \fI<displayname>\fB
Use the X-server display `displayname'.
.TP
.B \-\-session \fI<sessionId>\fB
Restore the application for the given `sessionId'.
.TP
.B \-\-cmap
Causes the application to install a private color map on an 8\-bit
display.
.TP
.B \-\-ncols \fI<count>\fB
Limits the number of colors allocated in the color cube on an 8\-bit
display, if the application is using the QApplication::ManyColor color
specification.
.TP
.B \-\-nograb
Tells Qt to never grab the mouse or the keyboard.
.TP
.B \-\-dograb
Running under a debugger can cause an implicit \-nograb, use \-dograb
to override.
.TP
.B \-\-sync
Switches to synchronous mode for debugging.
.TP
.B \-\-fn, \-\-font \fI<fontname>\fB
Defines the application font.
.TP
.B \-\-bg, \-\-background \fI<color>\fB
Sets the default background color and an application palette (light
and dark shades are calculated).
.TP
.B \-\-fg, \-\-foreground \fI<color>\fB
Sets the default foreground color.
.TP
.B \-\-btn, \-\-button \fI<color>\fB
Sets the default button color.
.TP
.B \-\-name \fI<name>\fB
Sets the application name.
.TP
.B \-\-title \fI<title>\fB
Sets the application title (caption).
.TP
.B \-\-visual TrueColor
Forces the application to use a TrueColor visual on an 8-bit display.
.TP
.B \-\-inputstyle \fI<inputstyle>\fB
Sets XIM (X Input Method) input style. Possible values are
\fIonthespot\fP, \fIoverthespot\fP, \fIoffthespot\fP and \fIroot\fP.
.TP
.B \-\-im \fI<XIM server>\fB
Set XIM server.
.TP
.B \-\-noxim
Disable XIM.

.SH TDE OPTIONS
.TP
.B \-\-caption \fI<caption>\fB
Use `caption' as name in the titlebar.
.TP
.B \-\-icon \fI<icon>\fB
Use `icon' as the application icon.
.TP
.B \-\-miniicon \fI<icon>\fB
Use `icon' as the icon in the titlebar.
.TP
.B \-\-dcopserver \fI<server>\fB
Use the DCOP Server specified by `server'.
.TP
.B \-\-nocrashhandler
Disable crash handler, to get core dumps.
.TP
.B \-\-waitforwm
Waits for a WM_NET compatible windowmanager.
.TP
.B \-\-style \fI<style>\fB
Sets the application GUI style.
.TP
.B \-\-geometry \fI<geometry>\fB
Sets the client geometry of the main widget.

.SH SEE ALSO
More detailed user documentation is available from
.BR help:/kmousetool
(either enter this URL into \fBkonqueror\fP, or run `khelpcenter
help:/kmousetool').

.PP
There is also further information available at 
.URL https://mirror.git.trinitydesktop.org/gitea/TDE/tdeaccessibility " 
tdeaccessibility project home page."

kmouthtool(1)

.SH AUTHORS
.nf
GunnarSchmi <gunnar@schmi>
.br
OlafSchmidt <ojschmidt@kde.org>
.br
JeffRoush <jeff@mousetool.com>


kttsd/ChangeLog
2006-11-02  Gary Cramblitt (PhantomsDad)
        * Add support for new Vietnamese voices to festivalintvoices file.  Get them at http://sourceforge.net/projects/vietnamesevoice.

2006-02-09  Gary Cramblitt (PhantomsDad)
        * Use KDE_CHECK_{HEADER,LIB} instead of AC_CHECK_{HEADER,LIB} to locate alsalib.

2006-02-07  Stephan Johach
        * Fix translation of "Speak Text" in kate/ktextedit.

2006-02-06  Gary Cramblitt (PhantomsDad)
        * Bump version to 0.3.5.2.
        * Fix BUG:121427 Translations sometimes not working due to incorrect catalog
          insertion/removal.

2006-01-20  ======== Release v0.3.5.1 with KDE 3.5.1

2006-01-13  Gary Cramblitt (PhantomsDad)
        * Fix BUG:120083 buffer length problem in Epos and Command plugin.
        * Fix BUG:115795 Don't try to start the Epos server for every utterance.
        * Epos server command changed from epos to eposd.  Epos client command changed from
          say to say-epos.  These changes appeared around Epos v2.5.35.  Enhance to try for these
          automatically.

2006-01-10  Gary Cramblitt (PhantomsDad)
        * Bug fix.  Pausing when popping up status in systray.
        * Plug small memory leak.
        * Allow user to specify custom ALSA device name.

2006-01-09  Gary Cramblitt (PhantomsDad)
        * Fix BUG:119753 Freezing when removing a speaking text job speaking on AlsaPlayer.
          AlsaPlayer now uses non-blocking pcm open and polling.
        * The ALSA device list pickable by user is now restricted to "default" and available
          "plughw" devices.
        * The following parameters may be placed in the [ALSAPlayer] section of
          ~/.kde/share/config/kttsdrc:
            PcmName=default    (Any valid alsa pcm device, but your on your own.  Examples:
                                dmix, hw:1,0)
            DebugLevel=1       (0 = errors only; 1 = messages; 2 = verbose debug)
            PeriodSize=128     (Number of frames in the ALSA buffer between interrupts.
                                If you have trouble, try adjusting up or down.)
            Periods=8          (Same)
        * If the ALSA pcm does not support pausing, simulates by not writing to the
          pcm anymore, which will cause a slight delay, and underruns,
          but that's the best we can do, I suppose.
        * Fix bug.  When starting KTTSMgr, it loaded the aRts plugin to determine if aRts is
          available.  But this started the aRts server and if user's machine does not have
          proper audio sharing enabled, caused "device busy" errors.  Now it starts the aRts
          server only when playback begins the first time.
        * Bump version from 0.3.5 to 0.3.5.1.

2006-01-06  Gary Cramblitt (PhantomsDad)
        * Fix BUG:118016 Not loading plugins when desktop language is not ISO-8859-1.

2005-12-03  Jürgen Zdero
        * Add German chat filter by Jürgen Zdero.

2005-11-13  Gary Cramblitt (PhantomsDad)
        * BUG:116031 Fix appendText method.  Add version method to dcop interface.
          This change did not make it into the KDE 3.5 release.

2005-11-29  ======== Release v0.3.5 with KDE 3.5

2005-10-16  J. Riddell
        * Make install error.  Install the Hadifix icons.

2005-07-19  Gary Cramblitt (PhantomsDad)
        * Name kept .wav files in format kttsd-jjjjjjjj-ssssssss.wav where jjjjjjjj is the job 
          number zero-filled to the left and ssssssss is the sentence sequence number zero-filled
          to the left.

2005-07-01  Gary Cramblitt (PhantomsDad)
        * Since ALSA plugin seems to work pretty well, enable building --without-arts and build
          ALSA plugin by default (if headers found).

2005-06-25  Gary Cramblitt (PhantomsDad)
        * The XML Transformer filter now applies itself when the root element OR doctype and appId match.

2005-06-12  Gary Cramblitt (PhantomsDad)
        * New aKode audio plugin.  Build it with ./configure --enable-kttsd-akode.
          There are a couple of problems.  See README players/akodeplayer/README.
        * Activate rich speak (XHTML to SSML) in Konqueror plugin.  To speak richly,
          you must configure the xhtml2ssml_simple XML Transformer plugin.

2005-06-03  Gary Cramblitt (PhantomsDad)
        * New ALSA audio plugin.  Build it with ./configure --enable-kttsd-alsa for now.
        * New Qt Rich Text to Plain Text filter by David Powell.

2005-04-30  Gary Cramblitt (PhantomsDad)
        * Change Current Sentence widget in Jobs tab to KTextEdit to avoid resizing when text does
          not fit in box.
        * HTML to SSML now working pretty well.  Note that you must
            1.  Have Festival 1.95 installed.
            2.  Have rab_diphone voice installed.
            3.  Have xsltproc installed.
            4.  If you have Festival Talker configured, reconfigure it (Edit) so that it will detect
                if you have rab_diphone voice installed.  You must hit the Apply button.
            5.  Enable XML Transformer filter and point it to xhtml2ssml_simple.xsl stylesheet.
            6.  Right now, only speaks HTML from Konqi if you paste HTML from clipboard.
                Speak button in Konqi not yet enabled until I get some confidence it is working well.

2005-04-23  Gary Cramblitt (PhantomsDad)
        * Break sentences longer than 600 characters into shorter sentences by changing comma to
          period.  This prevents Festival from droning on lower and lower in tone until it is
          unintelligible or crashes.

2005-04-22  Gary Cramblitt (PhantomsDad)
        * If resumeText is called on a speaking job, do not restart it.

2005-04-21  Gary Cramblitt (PhantomsDad)
        * New Hungarian Hadifix (mbrola and txt2pho) voice available at
            http://tkltrans.sourceforge.net/
          Since Hungarian requires ISO 8859-2 character encoding, added encoding option to
          Hadifix plugin.  Also, Hadifix plugin auto-selects a voice matching the user's
          chosen language.

kmousetool/kmousetool/kmousetool.cpp
/***************************************************************************
                          kmousetool.cpp  -  description
                             -------------------
    begin                : Sun Jan 20 23:27:58 PST 2002
    copyright            : (C) 2002-2003 by Jeff Roush
    email                : jeff@mousetool.com
    copyright            : (C) 2003 by Olaf Schmidt
    email                : ojschmidt@kde.org
    copyright            : (C) 2003 by Gunnar Schmi Dt
    email                : gunnar@schmi-dt.de
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/


#include "Xmd_kmousetool.h"
#include "kmousetool.h"
#include "kmousetool.moc"
#include <tdeconfig.h>
#include <X11/Intrinsic.h>     /* Intrinsics Definitions*/
#include <X11/StringDefs.h>    /* Standard Name-String definitions*/
#include <X11/extensions/xtestext1.h>    /* Standard Name-String definitions*/
#include <X11/extensions/XTest.h>    /* Standard Name-String definitions*/
#include <kdialog.h>
#include <tdelocale.h>
#include <tqpushbutton.h>
#include <tqpoint.h>
#include <tqnamespace.h>
#include <tdemessagebox.h>
#include <kaudioplayer.h>
#include <kstandarddirs.h>
#include <tqsound.h>
#include <tdeglobalsettings.h>
#include <kdebug.h>
#include <tqlayout.h>
#include <tqlineedit.h>
#include <tqcheckbox.h>
#include <ksystemtray.h>
#include <kiconloader.h>
#include <kpushbutton.h>
#include <kstdguiitem.h>
#include <knuminput.h>
#include <tdepopupmenu.h>
#include <tdeaboutapplication.h>

//#include <arts/soundserver.h>
#include <netwm.h>

#include <iostream>

#include "mtstroke.h"

//using namespace Arts;

int currentXPosition;
int currentYPosition;

//using namespace std;

#undef long

/**
* Gnarly X functions
*/
void queryPointer(Window *pidRoot, int *pnRx, int *pnRy, int *pnX, int *pnY, unsigned int *puMask);
int CursorHasMoved(int minMovement);
void LeftClick();
void RightClick();
void DoubleClick();
void LeftDn();
void LeftUp();

// Declarations
Display *display;

//SimpleSoundServer sound_server(Reference("global:Arts_SimpleSoundServer"));


void KMouseTool::init_vars()
{
	continue_timer = 1;
	tick_count = 0;
	max_ticks = dwell_time+1;
	mouse_is_down = FALSE;

	loadOptions();

	// If the ~/.mousetool directory doesn't exist, create it
//	mSoundFileName = TQDir::homeDirPath();
	mSoundFileName = locate("appdata", "sounds/mousetool_tap.wav");
	mplayer = new KAudioPlayer(mSoundFileName);


	// find application file
	appfilename = locate("exe", "kmousetool");

	// find the user's autostart directory
	autostartdirname = TDEGlobalSettings::autostartPath();

	// SimpleSoundServer server(Reference("global:Arts_SimpleSoundServer"));
//	sound_server(Reference("global:Arts_SimpleSoundServer"));

	TQDesktopWidget *d = TQApplication::desktop();
	int w = d->width();
	int h = d->height();
	MTStroke::setUpperLeft(0,0);
	MTStroke::setUpperRight(w-1,0);
	MTStroke::setLowerLeft(0,h-1);
	MTStroke::setLowerRight(w-1,h-1);
}

void KMouseTool::resetSettings()
{
	cbDrag ->setChecked(smart_drag_on);
	cbStart->setChecked(isAutostart());
	cbClick->setChecked(playSound);
	cbStroke->setChecked(strokesEnabled);
	movementEdit->setValue(min_movement);
	dwellTimeEdit->setValue(dwell_time);
	dragTimeEdit->setValue(drag_time);
	settingsChanged();
}

void KMouseTool::setDefaultSettings()
{
	cbDrag ->setChecked(false);
	cbStart->setChecked(false);
	cbClick->setChecked(false);
	cbStroke->setChecked(false);
	movementEdit->setValue(5);
	dwellTimeEdit->setValue(5);
	dragTimeEdit->setValue(3);
	settingsChanged();
}


void KMouseTool::timerEvent( TQTimerEvent * )
{
	if (!mousetool_is_running)
		return;

	if (!continue_timer) {
		TQT_TQOBJECT(this)->killTimers();
		return;
	}

	max_ticks = dwell_time + drag_time;
	stroke.addPt(currentXPosition, currentYPosition);

	moving = moving?CursorHasMoved(1):CursorHasMoved(min_movement);
	if (moving) {
		if (mousetool_just_started) {
			mousetool_just_started = false;
			tick_count = max_ticks;
		}
		else
			tick_count = 0;
		return;
	}

	if (tick_count<max_ticks)
		tick_count++;


	// If the mouse has paused ...
	if (tick_count==dwell_time) {
		int strokeType = stroke.getStrokeType();

		// if we're dragging the mouse, ignore stroke type
		if (mouse_is_down) {
			normalClick();
			return;
		}

		if (!strokesEnabled) {
			normalClick();
			return;
		}
		if (strokeType == MTStroke::DontClick)
			return;
		if (strokeType==MTStroke::bumped_mouse)
			return;

		if (strokeType==MTStroke::RightClick || strokeType==MTStroke::UpperRightStroke)
			RightClick();
		else if (strokeType==MTStroke::DoubleClick || strokeType==MTStroke::LowerLeftStroke)
			DoubleClick();
		else
			normalClick();
	}
}

void KMouseTool::normalClick()
{
	if (smart_drag_on) {
		if (!mouse_is_down) {
			LeftDn();
			mouse_is_down = TRUE;
			tick_count = 0;
			playTickSound();
		}
		else if (mouse_is_down) {
			LeftUp();
			mouse_is_down = FALSE;
			tick_count = max_ticks;
		}
	}
	else {
		// not smart_drag_on
		LeftClick();
		playTickSound();
	}
}

// This function isn't happy yet.
void KMouseTool::playTickSound()
{
	if (!playSound)
	return;

	mplayer->play();
	// This is a bit slow.
//	KAudioPlayer::play(mSoundFileName);
	return;

	// the solution seems to be to use MCOP,
	// but SimpleSoundServer is not cooperating

	// When using the following code, make sure to link with
	// Put " -lsoundserver_idl" in the linker options in KDevelop.

//  SimpleSoundServer server = SimpleSoundServer::_fromString("global:Arts_SimpleSoundServer");
//	SimpleSoundServer *ss = new SimpleSoundServer(Reference("global:Arts_SimpleSoundServer"));
//	SimpleSoundServer sound_server2(Reference("global:Arts_SimpleSoundServer"));

//	if(sound_server.isNull())
//		return;
//	sound_server.play(mSoundFileName.latin1());

	// Another approach is to try using TQSound.
	// TQSound depends on Network Audio Server, which doesn't seem to work on my Debian Woody system.
	// I haven't spent much time working on it, though.
//	TQSound::play(mSoundFileName);
}

KMouseTool::KMouseTool(TQWidget *parent, const char *name) : KMouseToolUI(parent, name)
{
	init_vars();
	resetSettings();

	connect(movementEdit, TQT_SIGNAL(valueChanged(int)), this, TQT_SLOT(settingsChanged()));
	connect(dwellTimeEdit, TQT_SIGNAL(valueChanged(int)), this, TQT_SLOT(settingsChanged()));
	connect(dragTimeEdit, TQT_SIGNAL(valueChanged(int)), this, TQT_SLOT(settingsChanged()));
	connect(cbDrag, TQT_SIGNAL(stateChanged(int)), this, TQT_SLOT(settingsChanged()));
	connect(cbStroke, TQT_SIGNAL(stateChanged(int)), this, TQT_SLOT(settingsChanged()));
	connect(cbClick, TQT_SIGNAL(stateChanged(int)), this, TQT_SLOT(settingsChanged()));
	connect(cbStart, TQT_SIGNAL(stateChanged(int)), this, TQT_SLOT(settingsChanged()));

	connect(buttonStartStop, TQT_SIGNAL(clicked()), this, TQT_SLOT(startStopSelected()));
	buttonDefault->setGuiItem(KStdGuiItem::defaults());
	connect(buttonDefault, TQT_SIGNAL(clicked()), this, TQT_SLOT(defaultSelected()));
	connect(buttonReset, TQT_SIGNAL(clicked()), this, TQT_SLOT(resetSelected()));
	buttonApply->setGuiItem(KStdGuiItem::apply());
	connect(buttonApply, TQT_SIGNAL(clicked()), this, TQT_SLOT(applySelected()));
	buttonHelp->setGuiItem(KStdGuiItem::help());
	connect(buttonHelp, TQT_SIGNAL(clicked()), this, TQT_SLOT(helpSelected()));
	buttonClose->setGuiItem(KStdGuiItem::close());
	connect(buttonClose, TQT_SIGNAL(clicked()), this, TQT_SLOT(closeSelected()));
#if TDE_VERSION >= TDE_MAKE_VERSION (3,1,90)
	buttonQuit->setGuiItem(KStdGuiItem::quit());
#endif // KDE 3.2
	connect(buttonQuit, TQT_SIGNAL(clicked()), this, TQT_SLOT(quitSelected()));

	// When we first start, it's nice to not click immediately.
	// So, tell MT we're just starting
	mousetool_just_started = true;

	startTimer(100);
	trayIcon = new KMouseToolTray (this, "systemTrayIcon");
	updateStartStopText ();
	connect(trayIcon, TQT_SIGNAL(startStopSelected()), this, TQT_SLOT(startStopSelected()));
	connect(trayIcon, TQT_SIGNAL(configureSelected()), this, TQT_SLOT(configureSelected()));
	connect(trayIcon, TQT_SIGNAL(aboutSelected()), this, TQT_SLOT(aboutSelected()));
	connect(trayIcon, TQT_SIGNAL(helpSelected()), this, TQT_SLOT(helpSelected()));
	connect(trayIcon, TQT_SIGNAL(quitSelected()), this, TQT_SLOT(quitSelected()));

	aboutDlg = new TDEAboutApplication (0, "KMouseToolDlg", false);
}

KMouseTool::~KMouseTool()
{
}

// *********************************************************
// Raw X functions
// Begin mouse monitoring and event generation code.
// these should be moved to a separate file.
void LeftClick()
{
	XTestFakeButtonEvent(display, 1, TRUE, 0);
	XTestFakeButtonEvent(display, 1, FALSE, 0);
}

void DoubleClick()
{
	XTestFakeButtonEvent(display, 1, TRUE, 0);
	XTestFakeButtonEvent(display, 1, FALSE, 100);
	XTestFakeButtonEvent(display, 1, TRUE, 200);
	XTestFakeButtonEvent(display, 1, FALSE, 300);
}

void RightClick()
{
	XTestFakeButtonEvent(display, 3, TRUE, 0);
	XTestFakeButtonEvent(display, 3, FALSE, 0);
}


void LeftDn()
{
	XTestFakeButtonEvent(display, 1, TRUE, 0);
}


void LeftUp()
{
	XTestFakeButtonEvent(display, 1, FALSE, 0);
}


void queryPointer(Window *pidRoot, int *pnRx, int *pnRy, int *pnX, int *pnY, unsigned int *puMask)
{
	Window id2, idRoot;
	int screen_num;

	screen_num = DefaultScreen(display);
	idRoot = RootWindow(display,screen_num);
	XQueryPointer(display, idRoot, &idRoot, &id2, pnRx, pnRy, pnX, pnY, puMask);

	*pidRoot = idRoot;
}

// return 1 if cursor has moved, 0 otherwise
int CursorHasMoved (int minMovement)
{
	static int nOldRootX = -1;
	static int nOldRootY = -1;

//	XEvent evButtonEvent;
	Window idRootWin;
	int nRootX, nRootY, nWinX, nWinY;
	unsigned int uintMask;

	queryPointer(&idRootWin, &nRootX, &nRootY, &nWinX, &nWinY, &uintMask);

	int nRes = 0;
	if ((nRootX>nOldRootX?nRootX-nOldRootX:nOldRootX-nRootX) >= minMovement)
		nRes = 1;
	if ((nRootY>nOldRootY?nRootY-nOldRootY:nOldRootY-nRootY) >= minMovement)
		nRes = 1;

	currentXPosition = nRootX;
	currentYPosition = nRootY;

	if (nRes) {
		nOldRootX = nRootX;
		nOldRootY = nRootY;
	}

	return nRes;
}
// End mouse monitoring and event creation code


bool KMouseTool::isAutostart()
{
	TQString sym = autostartdirname;
	sym += "kmousetool";			// sym is now full path to symlink
	TQFileInfo fi(sym);

	return fi.exists();
}

void KMouseTool::setAutostart (bool start)
{
	TQString sym = autostartdirname;
	sym += "kmousetool";			// sym is now full path to symlink
	TQFileInfo fi(sym);
	TQString cmd;

	if (start) {
		if (!fi.exists())  			// if it doesn't exist, make it
		cmd.sprintf("ln -s %s %s", appfilename.latin1(), autostartdirname.latin1());
	}
	else {
	if (fi.exists()) 			// if it exists, delete it
		cmd.sprintf("rm -f %s", sym.latin1());
	}
	system(cmd.ascii());
}

bool KMouseTool::applySettings()
{
	int drag, dwell;

	dwell = dwellTimeEdit->value();
	drag = dragTimeEdit->value() ;

	// The drag time must be less than the dwell time
	if ( dwell < drag) {
		KMessageBox::sorry(this, i18n("The drag time must be less than or equal to the dwell time."), i18n("Invalid Value"));
		return false;
	}

	// if we got here, we must be okay.
	min_movement   = movementEdit->value();
	smart_drag_on  = cbDrag->isChecked();
	playSound      = cbClick->isChecked();
	strokesEnabled = cbStroke->isChecked();
	setAutostart (cbStart->isChecked());

	dwell_time = dwell;
	drag_time  = drag;
	tick_count = max_ticks;

	saveOptions();
	settingsChanged();
	return true;
}

// Save options to kmousetoolrc file
void KMouseTool::loadOptions()
{
	kapp->config()->setGroup("UserOptions");

	playSound = kapp->config()->readBoolEntry("AudibleClick",false);
	smart_drag_on = kapp->config()->readBoolEntry("SmartDrag",false);

	dwell_time = kapp->config()->readNumEntry("DwellTime",5);
	drag_time = kapp->config()->readNumEntry("DragTime",3);
	min_movement = kapp->config()->readNumEntry("Movement",5);
	strokesEnabled = kapp->config()->readBoolEntry("strokesEnabled",false);

	TQPoint p;
	int x = kapp->config()->readNumEntry("x",0);
	int y = kapp->config()->readNumEntry("y",0);
	p.setX(x);
	p.setY(y);
	move(p);

	mousetool_is_running = kapp->config()->readBoolEntry("MouseToolIsRunning",false);
	display = XOpenDisplay(NULL);
}

// Save options to kmousetoolrc file
void KMouseTool::saveOptions()
{
	TQPoint p = pos();
	int x = p.x();
	int y = p.y();

	kapp->config()->setGroup("ProgramOptions");
	kapp->config()->writeEntry("Version", KMOUSETOOL_VERSION);
	kapp->config()->setGroup("UserOptions");
	kapp->config()->writeEntry("x", x);
	kapp->config()->writeEntry("y", y);
	kapp->config()->writeEntry("strokesEnabled", strokesEnabled);
	kapp->config()->writeEntry("IsMinimized", isHidden());
	kapp->config()->writeEntry("DwellTime", dwell_time);
	kapp->config()->writeEntry("DragTime", drag_time);
	kapp->config()->writeEntry("Movement", min_movement);
	kapp->config()->writeEntry("SmartDrag", smart_drag_on);
	kapp->config()->writeEntry("AudibleClick", playSound);
	kapp->config()->writeEntry("MouseToolIsRunning", mousetool_is_running);
	kapp->config()->sync();
}

void KMouseTool::updateStartStopText()
{
	if (mousetool_is_running)
		buttonStartStop->setText(i18n("&Stop"));
	else
		buttonStartStop->setText(i18n("&Start"));
	trayIcon->updateStartStopText(mousetool_is_running);
}

bool KMouseTool::newSettings()
{
	return ((dwell_time != dwellTimeEdit->value()) ||
		(drag_time != dragTimeEdit->value()) ||
		(min_movement != movementEdit->value()) ||
		(smart_drag_on != cbDrag->isChecked()) ||
		(playSound != cbClick->isChecked()) ||
		(strokesEnabled != cbStroke->isChecked()) ||
		(isAutostart() != cbStart->isChecked()));
}

bool KMouseTool::defaultSettings()
{
	return ((5 == dwellTimeEdit->value()) &&
		(3 == dragTimeEdit->value()) &&
		(5 == movementEdit->value()) &&
		!cbDrag->isChecked() &&
		!cbClick->isChecked() &&
		!cbStroke->isChecked() &&
		!cbStart->isChecked());
}

/******** SLOTS **********/

// Value or state changed
void KMouseTool::settingsChanged ()
{
	buttonReset->setEnabled (newSettings());
	buttonApply->setEnabled (newSettings());
	buttonDefault->setDisabled (defaultSettings());
}

// Buttons within the dialog
void KMouseTool::startStopSelected()
{
	mousetool_is_running = !mousetool_is_running;
	updateStartStopText();
}

void KMouseTool::defaultSelected()
{
	setDefaultSettings();
}

void KMouseTool::resetSelected()
{
	resetSettings();
}

void KMouseTool::applySelected()
{
	applySettings();
}

// Buttons at the bottom of the dialog
void KMouseTool::helpSelected()
{
	kapp->invokeHelp();
}

void KMouseTool::closeSelected()
{
	if (newSettings())
	{
		int answer = KMessageBox::questionYesNoCancel (this,
			i18n("There are unsaved changes in the active module.\nDo you want to apply the changes before closing the configuration window or discard the changes?"),
			i18n("Closing Configuration Window"),
			KStdGuiItem::apply(), KStdGuiItem::discard(), "AutomaticSave");
		if (answer == KMessageBox::Yes)
			applySettings();
		else if (answer == KMessageBox::No)
			resetSettings();
		if (answer != KMessageBox::Cancel)
			hide();
	}
	else
		hide();
}

void KMouseTool::quitSelected()
{
	if (newSettings())
	{
		int answer = KMessageBox::questionYesNoCancel (this,
			i18n("There are unsaved changes in the active module.\nDo you want to apply the changes before quitting KMousetool or discard the changes?"),
			i18n("Quitting KMousetool"),
			KStdGuiItem::apply(), KStdGuiItem::discard(), "AutomaticSave");
		if (answer == KMessageBox::Yes)
			applySettings();
		if (answer != KMessageBox::Cancel)
		{
			saveOptions();
			kapp->quit();
		}
	}
	else
	{
		saveOptions();
		kapp->quit();
	}
}

// Menu functions
void KMouseTool::configureSelected()
{
	show();
	raise();
	setActiveWindow();
}

void KMouseTool::aboutSelected()
{
	aboutDlg->show();
}



KMouseToolTray::KMouseToolTray (TQWidget *parent, const char *name) : KSystemTray (parent, name)
{
	startStopId = contextMenu()->insertItem (i18n("&Start"), this, TQT_SIGNAL(startStopSelected()));
	contextMenu()->insertSeparator();
	contextMenu()->insertItem (TDEGlobal::iconLoader()->loadIcon("configure", TDEIcon::Small),
	                           i18n("&Configure KMouseTool..."), this, TQT_SIGNAL(configureSelected()));
	contextMenu()->insertSeparator();
	contextMenu()->insertItem (TDEGlobal::iconLoader()->loadIcon("contents", TDEIcon::Small),
	                           i18n("KMousetool &Handbook"), this, TQT_SIGNAL(helpSelected()));
	contextMenu()->insertItem (TDEGlobal::iconLoader()->loadIcon("kmousetool", TDEIcon::Small),
	                           i18n("&About KMouseTool"), this, TQT_SIGNAL(aboutSelected()));
}

KMouseToolTray::~KMouseToolTray() {
}

void KMouseToolTray::updateStartStopText(bool mousetool_is_running)
{
	TQPixmap icon;

	if (mousetool_is_running) {
		contextMenu()->changeItem (startStopId, i18n("&Stop"));
		icon = TDEGlobal::iconLoader()->loadIcon("kmousetool_on", TDEIcon::Small);
	}
	else {
		contextMenu()->changeItem (startStopId, i18n("&Start"));
		icon = TDEGlobal::iconLoader()->loadIcon("kmousetool_off", TDEIcon::Small);
	}
	setPixmap (icon);
	show();
}
